File: com/white/GenerateParentheses.java
package com.white;

import java.util.ArrayList;
import java.util.List;

public class GenerateParentheses {

    // 22. Generate Parentheses
    public static List<String> generateParenthesis(int n) {
        
        List<String> result = new ArrayList<String>();
        if (n == 0) {
            return result;
        }
        helper(result, "", n, n);
        return result;
    }

    // Amazing solution, need to learn more about it
    private static void helper(List<String> res, String present, int left, int right) {
        if (right == 0) {
            res.add(present);
        }
        if (left > 0) {
            helper(res, present + "(", left - 1, right);
        }
        if (right > left) {
            helper(res, present + ")", left, right - 1);
        }
    }
}

File: LeetcodeCore/GenerateParentheses.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LeetcodeCore
{
    public class GenerateParentheses
    {
        // 22. Generate Parentheses
        public IList<string> GenerateParenthesis(int n)
        {
            var results = new List<string>();
            Backtracking(results, "", n, 0, 0);
            return results;
        }

        private void Backtracking(IList<string> results, string s, int n, int left, int right)
        {
            if (s.Length == n*2)
            {
                results.Add(s);
                return;
            }

            if (left < n)
                Backtracking(results, s + "(", n, left + 1, right);
            if (right < left)
                Backtracking(results, s + ")", n, left, right + 1);
        }
    }
}
